### PR TITLE
add a "parse plugin" mechanism to OpalCompiler

### PR DESCRIPTION
**Pull request by estebanlm**
this is o be used to transform the ast in the parse phase (otherwise plugins are used in compile phase).
we will use this mechanism to add custom highlighter.

### CODE REVIEW

File: src/OpalCompiler-Core/CompilationContext.class.st [hunk 1]
```diff
@@ -466,7 +481,8 @@ CompilationContext >> getClass [
 CompilationContext >> initialize [
 	options := Set new.
 	astTransformPlugins := OrderedCollection withAll: self class defaultTransformationPlugins.
-	semanticScope := OCMethodSemanticScope targetingClass: nil class
+	semanticScope := OCMethodSemanticScope targetingClass: nil class.
+	astParseTransformPlugins := OrderedCollection new: 2
```
Comment (privat) [inline src/OpalCompiler-Core/CompilationContext.class.st]: introduce `defaultParseTransformationPlugins` to be consistent with the other plugins?

**Review by privat — APPROVED** [review]
Did you consider renaming the other plugin? E.g. from `astTransformPlugins` to `astCompileTransformPlugins` etc.

File: src/OpalCompiler-Core/CompilationContext.class.st (hunk 1 above)
Comment (guillep) [inline src/OpalCompiler-Core/CompilationContext.class.st]: Is it really needed to allocate it with 2?

File: src/AST-Core/RBMessageNode.class.st [hunk 2]
```diff
@@ -474,6 +474,14 @@ RBMessageNode >> sentMessages [
 		yourself
 ]
 
+{ #category : #accessing }
+RBMessageNode >> start: newStart [
+	| oldStart |
+
+	oldStart := self start.
+	keywordsPositions := keywordsPositions collect: [ :each | each - oldStart + newStart ]
```
Comment (guillep) [inline src/AST-Core/RBMessageNode.class.st]: This could use a test

File: src/OpalCompiler-Core/OpalCompiler.class.st [hunk 3]
```diff
@@ -199,12 +205,22 @@ OpalCompiler >> bindings: aDictionary [
...
+
+	plugins := compilationContext astParseTransformPlugins ifEmpty: [ ^ self ].
+	plugins sort: [ :a :b | a priority > b priority ]. "priority 0 is sorted last"
+	plugins do: [ :each | ast := each transform: ast ]
+]
+
 { #category : #private }
 OpalCompiler >> callPlugins [
-	| plugins  |
+	| plugins |
```
Comment (guillep) [inline src/OpalCompiler-Core/OpalCompiler.class.st:219]: Now that we have some other plugins maybe we need to rename this? And keep of course the following for backwards compatibility

```smalltalk
callPlugins
   self callCompileTimePlugins
```

File: src/Shout/Behavior.extension.st [hunk 4]
```diff
@@ -0,0 +1,10 @@
+Extension { #name : #Behavior }
+
+{ #category : #'*Shout' }
+Behavior >> syntaxHighlighter [
```
Comment (guillep) [inline src/Shout/Behavior.extension.st]: I don't think this should be in the PR, right?

Unless the PR is about pluggable highlighting, not about parse plugin mechanism.

File: src/Shout/SHRBTextStyler.class.st [hunk 5]
```diff
@@ -14,7 +14,8 @@ Class {
 		'font',
 		'isForWorkspace',
 		'pixelHeight',
-		'workspace'
+		'workspace',
+		'plugins'
```
Comment (guillep) [inline src/Shout/SHRBTextStyler.class.st]: Same as above for all changes in this file.
I don't think this should be in the PR, right?

Unless the PR is about pluggable highlighting, not about parse plugin mechanism.

**Review by guillep — CHANGES_REQUESTED** [review]

File: src/OpalCompiler-Core/CompilationContext.class.st (hunk 1 above)
Comment (estebanlm) [reply src/OpalCompiler-Core/CompilationContext.class.st]: ok

Comment (estebanlm) [reply src/OpalCompiler-Core/CompilationContext.class.st]: it is a size optimization: most of the time there will not be parse plugin so allocate 10 slots for it seemed a lot, but anyway, that's already addressed.

File: src/Shout/Behavior.extension.st (hunk 4 above)
Comment (estebanlm) [reply src/Shout/Behavior.extension.st]: is about both. One is the proof/first use of the other.

**Comment by estebanlm** [comment]
> Did you consider renaming the other plugin? E.g. from `astTransformPlugins` to `astCompileTransformPlugins` etc.

not really, this is out of my scope :)
I will let @MarcusDenker to lead that :P

File: src/Shout/Behavior.extension.st (hunk 4 above)
Comment (guillep) [reply src/Shout/Behavior.extension.st]: Then changing the PR title was enough.

I don't understand why the commit title makes me happy, personally, it does not make me happy.

**Comment by estebanlm** [comment]
so, can this be taken care, please?
@MarcusDenker ?

File: src/OpalCompiler-Core/OpalCompiler.class.st (hunk 3 above)
Comment (MarcusDenker) [reply src/OpalCompiler-Core/OpalCompiler.class.st:219]: yes, we can improve naming im a next step.

**Review by MarcusDenker — APPROVED** [review]

**Pull request merged**